mu/virtual_environment.py

```diff
@@ -24,6 +25,40 @@
 logger = logging.getLogger(__name__)
 
 
+class SplashLogHandler(logging.NullHandler):
+    """
+    A simple log handler that does only one thing: use the referenced Qt signal
+    to emit the log.
+    """
+
+    def __init__(self, emitter):
+        """
+        Returns an instance of the class that will use the Qt signal passed in
+        as emitter.
+        """
+        super().__init__()
+        self.setLevel(logging.DEBUG)
+        self.emitter = emitter
+
+    def emit(self, record):
+        """
+        Emits a record via the Qt signal.
+        """
+        timestamp = datetime.datetime.fromtimestamp(record.created)
+        messages = record.getMessage().splitlines()
+        for msg in messages:
+            output = "[{level}]({timestamp}) - {message}".format(
+                level=record.levelname, timestamp=timestamp, message=msg
+            )

```

Comment:
Not 100% sure how much data we get in a `record`, but if it's a single call to the logger, shoudn't the entire message have a single timestamp, instead of adding a timestamp for each line of the same message?

Reply:
So, I treat each line as a unique log entry (even if it's a part of a multi-line message). This is so we only ever display n number of lines on the splash screen and text doesn't bleed into the bopping snake. ;-) That's why they all have the timestamp... but I agree that this timestamp should probably only be calculated once and re-used for all lines in a single message.